File: src/stack.rs
use super::frame::{Frame, Source};
use super::types::*;
use log::info;
use std::rc::Rc;

#[derive(Debug)]
pub struct Stack {
    pub frames: Vec<Frame>,
}

impl Stack {
    pub fn new(source: Source) -> Self {
        info!("{} | Initial frame {:?}", 0, source);
        Stack {
            frames: vec![Frame::new(source, 0)],
        }
    }

    pub fn get_vbl(&mut self, sym: &Symbol, usage: usize) -> Rc<Value> {
        // if this is the final usage, then pluck it out.
        let idx = match self.frames[0]
            .bindings
            .iter()
            .position(|(a, _, _)| a == sym)
        {
            None => unreachable!("Variable {:?} not found!", sym),
            Some(idx) => idx,
        };
        // TODO take usage into account
        if self.frames[0].bindings[idx].1 == usage {
            let (_, _, v) = self.frames[0].bindings.remove(idx);
            v
        } else {
            self.frames[0].bindings[idx].2.clone()
        }
    }

    pub fn new_frame(&mut self, return_index: usize, source: Source) {
        info!("{} | ----> New frame {:?}", self.frames.len(), source);
        self.frames.insert(0, Frame::new(source, return_index))
    }

    pub fn clone_frame(&mut self, return_index: usize) {
        info!("{} | ----> Clone frame", self.frames.len());
        self.frames.insert(0, self.frames[0].clone());
        self.frames[0].return_index = return_index;
    }

    // TODO there should be a way to ... get back .. to the thing ... that we wanted ...
    pub fn back_again_to_handler(
        &mut self,
        frames: &Vec<Frame>,
        current_idx: usize,
    ) -> Option<(usize, usize)> {
        // let mut frames = vec![];
        // self.frames.remove(0); // ignore the current one, it was a clone anyway
        let mut new_idx = current_idx + 1;
        while frames[new_idx].handler == None {
            new_idx += 1;
        }
        self.frames = frames[new_idx..].to_vec();
        // while self.frames[0].handler == None {
        //     frames.push(self.frames.remove(0));
        //     if self.frames.len() == 0 {
        //         return None;
        //     }
        // }
        let idx = self.frames[0].handler.unwrap();
        self.frames[0].handler = None;
        Some((idx, new_idx))
    }

    // TODO there should be a way to ... get back .. to the thing ... that we wanted ...
    pub fn back_to_handler(&mut self) -> Option<(usize, Vec<Frame>, usize)> {
        let mut frames = vec![];
        while self.frames[0].handler == None {
            frames.push(self.frames.remove(0));
            if self.frames.len() == 0 {
                return None;
            }
        }
        let current_idx = frames.len() - 1;
        // We need the full stack, in case this gets rethrown.
        // PERF: we could check up the stack and only do this if
        // there's another handler somewhere.
        frames.extend(self.frames.clone());
        let idx = self.frames[0].handler.unwrap();
        self.frames[0].handler = None;
        Some((idx, frames, current_idx))
    }

    pub fn pop_frame(&mut self) -> (usize, Rc<Value>) {
        let idx = self.frames[0].return_index;
        let value = self.pop().unwrap();
        self.frames.remove(0);
        info!(
            "{} | <---- Return to idx {} with value {:?} - {:?}",
            self.frames.len(),
            idx,
            value,
            self.frames[0].source
        );
        (idx, value)
    }
    // TODO : fn replace_frame
    pub fn push(&mut self, t: Rc<Value>) {
        // info!("{} | Stack push: {:?}", self.frames.len(), t);
        self.frames[0].stack.push(t);
    }
    pub fn pop(&mut self) -> Option<Rc<Value>> {
        let t = self.frames[0].stack.pop();
        // info!("{} | Stack pop: {:?}", self.frames.len(), t);
        t
    }
    // TODO maybe return a & ref to the Rc?
    pub fn peek(&mut self) -> Option<Rc<Value>> {
        let l = self.frames[0].stack.len();
        if l > 0 {
            // info!("Stack peek: {:?}", self.frames[0].stack[l - 1]);
            Some(self.frames[0].stack[l - 1].clone())
        } else {
            None
        }
    }
    pub fn pop_to_mark(&mut self) {
        let mark = self.frames[0].marks.pop().unwrap();
        while self.frames[0].stack.len() > mark {
            self.frames[0].stack.pop();
        }
    }
    pub fn mark(&mut self) {
        let ln = self.frames[0].stack.len();
        self.frames[0].marks.push(ln);
    }
    pub fn clear_mark(&mut self) {
        self.frames[0].marks.pop();
    }
    pub fn pop_up(&mut self) {
        let ln = self.frames[0].stack.len();
        self.frames[0].stack.remove(ln - 2);
    }
}

File: js/stack.js
const newFrame = (source, return_index) => ({
    source: source,
    stack: [],
    marks: [],
    handler: null,
    return_index: return_index,
    bindings: [],
    // traceId,
});

const symEq = (a, b) => {
    return a.unique === b.unique;
};

const showSource = (source) => {
    if (source.type === 'Fn') {
        return `${source.hash.slice(0, 10)} : Fn(${source.fnid})`;
    }
    if (source.type === 'term') {
        return `${source.hash.slice(0, 10)} : Value`;
    }
    return source;
};

const DEBUG = true;

export class Stack {
    constructor(source) {
        this.traceCounter = 0;
        this._frames = [newFrame(source, 0)];
    }

    bind(binding) {
        this._frames[0].bindings.unshift(binding);
    }

    bindLast(binding) {
        this._frames[0].bindings.push(binding);
    }

    currentFrame() {
        return this._frames[0];
    }

    isLastFrame() {
        return this._frames.length <= 1;
    }

    resumeContinuation(frames, returnIdx) {
        frames = JSON.parse(JSON.stringify(frames));
        const last = frames.length - 1;
        frames[last].return_index = returnIdx;
        frames.push(...this._frames);
        this._frames = frames;
    }

    get_vbl(sym, usage) {
        const idx = this._frames[0].bindings.findIndex((b) => symEq(b[0], sym));
        if (idx === -1) {
            throw new Error('sym not found');
        }
        const item = this._frames[0].bindings[idx];
        if (item[1] === usage) {
            this._frames[0].bindings.splice(idx, 1);
            return item[2];
        }
        return item[2];
    }

    new_frame(return_index, source) {
        if (DEBUG) {
            console.log(
                `->> ${this._frames.length} New Frame : `,
                showSource(source),
                return_index,
            );
        }
        this._frames.unshift(newFrame(source, return_index));
    }

    clone_frame(return_index) {
        if (DEBUG) {
            console.log('->> Clone Frame');
        }
        this._frames.unshift({ ...this._frames[0] });
        this._frames[0].return_index = return_index;
    }

    back_again_to_handler(frames, current_idx) {
        let new_idx = current_idx + 1;
        while (frames[new_idx].handler === null) {
            new_idx += 1;
        }
        this._frames = frames.slice(new_idx);
        let idx = this._frames[0].handler;
        if (idx == null) {
            throw new Error('no handler');
        }
        this._frames[0].handler = null;
        return [idx, new_idx];
    }

    back_to_handler() {
        const frames = [];
        while (this._frames[0].handler === null) {
            frames.push(this._frames.shift());
            if (this._frames.length === 0) {
                return null;
            }
        }
        const current_idx = frames.length - 1;
        frames.push(...this._frames.map((f) => JSON.parse(JSON.stringify(f))));
        const idx = this._frames[0].handler;
        if (idx == null) {
            throw new Error('no handler');
        }
        this._frames[0].handler = null;
        return [idx, frames, current_idx];
    }

    pop_frame() {
        const idx = this._frames[0].return_index;
        const value = this.pop();
        if (DEBUG) {
            console.log(
                `<<- ${this._frames.length} Pop Frame : ${idx} - ${showSource(
                    this._frames[1].source,
                )}`,
                value,
            );
        }
        if (value == null) {
            throw new Error('no return value');
        }
        this._frames.shift();
        return [idx, value];
    }

    push(t) {
        this._frames[0].stack.push(t);
        if (DEBUG) {
            console.log(
                'push to stack',
                t,
                'stack size',
                this._frames[0].stack.length,
            );
        }
    }

    pop() {
        const t = this._frames[0].stack.pop();
        if (DEBUG) {
            console.log('pop from stack', t);
        }
        if (t == null) {
            console.log(this._frames);
            throw new Error(`Popping from the stack, but nothing is there`);
        }
        return t;
    }

    peek() {
        const l = this._frames[0].stack.length;
        if (l > 0) {
            return this._frames[0].stack[l - 1];
        }
        return null;
    }

    pop_to_mark() {
        const mark = this._frames[0].marks.pop();
        if (mark == null) {
            throw new Error('pop to mark');
        }
        while (this._frames[0].stack.length > mark) {
            this._frames[0].stack.pop();
        }
    }

    mark() {
        const ln = this._frames[0].stack.length;
        this._frames[0].marks.push(ln);
    }

    clear_mark() {
        this._frames[0].marks.pop();
    }

    pop_up() {
        const ln = this._frames[0].stack.length;
        this._frames[0].stack.splice(ln - 2, 1);
    }
}
